File: fotoapparat/src/main/java/io/fotoapparat/FotoapparatBuilder.java
package io.fotoapparat;

import android.content.Context;
import android.support.annotation.IntRange;
import android.support.annotation.NonNull;

import io.fotoapparat.error.CameraErrorCallback;
import java.util.Collection;

import io.fotoapparat.hardware.CameraDevice;
import io.fotoapparat.hardware.provider.CameraProvider;
import io.fotoapparat.log.Logger;
import io.fotoapparat.log.Loggers;
import io.fotoapparat.parameter.Flash;
import io.fotoapparat.parameter.FocusMode;
import io.fotoapparat.parameter.LensPosition;
import io.fotoapparat.parameter.ScaleType;
import io.fotoapparat.parameter.Size;
import io.fotoapparat.parameter.range.Range;
import io.fotoapparat.parameter.selector.FlashSelectors;
import io.fotoapparat.parameter.selector.SelectorFunction;
import io.fotoapparat.parameter.selector.Selectors;
import io.fotoapparat.preview.FrameProcessor;
import io.fotoapparat.view.CameraRenderer;
import io.fotoapparat.view.CameraView;

import static io.fotoapparat.hardware.provider.CameraProviders.v1;
import static io.fotoapparat.parameter.selector.FocusModeSelectors.autoFocus;
import static io.fotoapparat.parameter.selector.FocusModeSelectors.continuousFocus;
import static io.fotoapparat.parameter.selector.FocusModeSelectors.fixed;
import static io.fotoapparat.parameter.selector.LensPositionSelectors.back;
import static io.fotoapparat.parameter.selector.LensPositionSelectors.external;
import static io.fotoapparat.parameter.selector.LensPositionSelectors.front;
import static io.fotoapparat.parameter.selector.Selectors.firstAvailable;
import static io.fotoapparat.parameter.selector.SizeSelectors.biggestSize;

/**
 * Builder for {@link Fotoapparat}.
 */
public class FotoapparatBuilder {

    Context context;
    CameraProvider cameraProvider = v1();
    CameraRenderer renderer;

    SelectorFunction<Collection<LensPosition>, LensPosition> lensPositionSelector = firstAvailable(
            back(),
            front(),
            external()
    );
    SelectorFunction<Collection<Size>, Size> photoSizeSelector = biggestSize();
    SelectorFunction<Collection<Size>, Size> previewSizeSelector = biggestSize();
    SelectorFunction<Collection<FocusMode>, FocusMode> focusModeSelector = firstAvailable(
            continuousFocus(),
            autoFocus(),
            fixed()
    );
    SelectorFunction<Collection<Flash>, Flash> flashSelector = FlashSelectors.off();
    SelectorFunction<Collection<Range<Integer>>, Range<Integer>> previewFpsRangeSelector = Selectors.nothing();
    SelectorFunction<Range<Integer>, Integer> sensorSensitivitySelector = Selectors.nothing();

    Integer jpegQuality;

    ScaleType scaleType = ScaleType.CENTER_CROP;

    FrameProcessor frameProcessor = null;

    Logger logger = Loggers.none();

    CameraErrorCallback cameraErrorCallback = CameraErrorCallback.NULL;

    FotoapparatBuilder(@NonNull Context context) {
        this.context = context;
    }

    /**
     * @param cameraProvider decides which {@link CameraDevice} to use.
     */
    public FotoapparatBuilder cameraProvider(@NonNull CameraProvider cameraProvider) {
        this.cameraProvider = cameraProvider;
        return this;
    }

    /**
     * @param selector selects size of the photo (in pixels) from list of available sizes.
     */
    public FotoapparatBuilder photoSize(@NonNull SelectorFunction<Collection<Size>, Size> selector) {
        photoSizeSelector = selector;
        return this;
    }

    /**
     * @param selector selects size of preview stream (in pixels) from list of available sizes.
     */
    public FotoapparatBuilder previewSize(@NonNull SelectorFunction<Collection<Size>, Size> selector) {
        previewSizeSelector = selector;
        return this;
    }

    /**
     * @param scaleType of preview inside the view.
     */
    public FotoapparatBuilder previewScaleType(ScaleType scaleType) {
        this.scaleType = scaleType;
        return this;
    }

    /**
     * @param selector selects focus mode from list of available modes.
     */
    public FotoapparatBuilder focusMode(@NonNull SelectorFunction<Collection<FocusMode>, FocusMode> selector) {
        focusModeSelector = selector;
        return this;
    }

    /**
     * @param selector selects flash mode from list of available modes.
     */
    public FotoapparatBuilder flash(@NonNull SelectorFunction<Collection<Flash>, Flash> selector) {
        flashSelector = selector;
        return this;
    }

    /**
     * @param selector camera sensor position from list of available positions.
     */
    public FotoapparatBuilder lensPosition(@NonNull SelectorFunction<Collection<LensPosition>, LensPosition> selector) {
        lensPositionSelector = selector;
        return this;
    }

    /**
     * @param selector selects preview FPS range from list of available ranges.
     */
    public FotoapparatBuilder previewFpsRange(@NonNull SelectorFunction<Collection<Range<Integer>>, Range<Integer>> selector) {
        previewFpsRangeSelector = selector;
        return this;
    }

    /**
     * @param selector selects ISO value from range of available values.
     */
    public FotoapparatBuilder sensorSensitivity(@NonNull SelectorFunction<Range<Integer>, Integer> selector) {
        sensorSensitivitySelector = selector;
        return this;
    }

    /**
     * @param jpegQuality of the picture (1-100)
     */
    public FotoapparatBuilder jpegQuality(@IntRange(from=0,to=100) @NonNull Integer jpegQuality) {
        this.jpegQuality = jpegQuality;
        return this;
    }

    /**
     * @param frameProcessor receives preview frames for processing.
     * @see FrameProcessor
     */
    public FotoapparatBuilder frameProcessor(@NonNull FrameProcessor frameProcessor) {
        this.frameProcessor = frameProcessor;
        return this;
    }

    /**
     * @param logger logger which will print logs. No logger is set by default.
     * @see Loggers
     */
    public FotoapparatBuilder logger(@NonNull Logger logger) {
        this.logger = logger;
        return this;
    }

    /**
     * @param callback which will be notified when camera error happens in Fotoapparat.
     * @see CameraErrorCallback
     */
    public FotoapparatBuilder cameraErrorCallback(@NonNull CameraErrorCallback callback) {
        this.cameraErrorCallback = callback;
        return this;
    }

    /**
     * @param renderer view which will draw the stream from the camera.
     * @see CameraView
     */
    public FotoapparatBuilder into(@NonNull CameraRenderer renderer) {
        this.renderer = renderer;
        return this;
    }

    /**
     * @return set up instance of {@link Fotoapparat}.
     * @throws IllegalStateException if some mandatory parameters are not specified.
     */
    public Fotoapparat build() {
        validate();

        return Fotoapparat.create(this);
    }

    private void validate() {
        if (cameraProvider == null) {
            throw new IllegalStateException("CameraProvider is mandatory.");
        }

        if (renderer == null) {
            throw new IllegalStateException("CameraRenderer is mandatory.");
        }

        if (lensPositionSelector == null) {
            throw new IllegalStateException("LensPosition selector is mandatory.");
        }

        if (photoSizeSelector == null) {
            throw new IllegalStateException("Photo size selector is mandatory.");
        }
    }

}

File: fotoapparat/src/main/java/io/fotoapparat/FotoapparatBuilder.kt
package io.fotoapparat

import android.content.Context
import io.fotoapparat.configuration.CameraConfiguration
import io.fotoapparat.error.CameraErrorCallback
import io.fotoapparat.error.CameraErrorListener
import io.fotoapparat.log.Logger
import io.fotoapparat.log.none
import io.fotoapparat.parameter.ScaleType
import io.fotoapparat.preview.Frame
import io.fotoapparat.selector.*
import io.fotoapparat.util.FrameProcessor
import io.fotoapparat.view.CameraRenderer
import io.fotoapparat.view.CameraView
import io.fotoapparat.view.FocusView
import io.fotoapparat.preview.FrameProcessor as FrameProcessorJava

/**
 * Builder for [Fotoapparat].
 */
class FotoapparatBuilder internal constructor(private var context: Context) {

    internal var lensPositionSelector: LensPositionSelector = firstAvailable(
            back(),
            front(),
            external()
    )
    internal var cameraErrorCallback: CameraErrorCallback = {}
    internal var renderer: CameraRenderer? = null
    internal var focusView: FocusView? = null
    internal var scaleType: ScaleType = ScaleType.CenterCrop
    internal var logger: Logger = none()

    internal var configuration = CameraConfiguration.default()

    /**
     * @param selector camera sensor position from list of available positions.
     */
    fun lensPosition(selector: LensPositionSelector): FotoapparatBuilder =
            apply { lensPositionSelector = selector }

    /**
     * @param scaleType of preview inside the view.
     */
    fun previewScaleType(scaleType: ScaleType): FotoapparatBuilder =
            apply { this.scaleType = scaleType }

    /**
     * @param selector selects resolution of the photo (in pixels) from list of available resolutions.
     */
    fun photoResolution(selector: ResolutionSelector): FotoapparatBuilder = apply {
        configuration = configuration.copy(
                pictureResolution = selector
        )
    }

    /**
     * @param selector selects size of preview stream (in pixels) from list of available resolutions.
     */
    fun previewResolution(selector: ResolutionSelector): FotoapparatBuilder = apply {
        configuration = configuration.copy(
                previewResolution = selector
        )
    }


    /**
     * @param selector selects focus mode from list of available modes.
     */
    fun focusMode(selector: FocusModeSelector): FotoapparatBuilder = apply {
        configuration = configuration.copy(
                focusMode = selector
        )
    }

    /**
     * @param selector selects flash mode from list of available modes.
     */
    fun flash(selector: FlashSelector): FotoapparatBuilder = apply {
        configuration = configuration.copy(
                flashMode = selector
        )
    }

    /**
     * @param selector selects preview FPS range from list of available ranges.
     */
    fun previewFpsRange(selector: FpsRangeSelector): FotoapparatBuilder = apply {
        configuration = configuration.copy(
                previewFpsRange = selector
        )
    }

    /**
     * @param selector selects ISO value from range of available values.
     */
    fun sensorSensitivity(selector: SensorSensitivitySelector): FotoapparatBuilder = apply {
        configuration = configuration.copy(
                sensorSensitivity = selector
        )
    }

    /**
     * @param selector of the Jpeg picture quality.
     */
    fun jpegQuality(selector: QualitySelector): FotoapparatBuilder = apply {
        configuration = configuration.copy(
                jpegQuality = selector
        )
    }

    /**
     * @param selector selects exposure compensation value from available range.
     */
    fun exposureCompensation(selector: ExposureSelector): FotoapparatBuilder = apply {
        configuration = configuration.copy(
                exposureCompensation = selector
        )
    }

    /**
     * @param frameProcessor receives preview frames for processing.
     * @see FrameProcessorJava
     */
    fun frameProcessor(frameProcessor: FrameProcessor): FotoapparatBuilder = apply {
        configuration = configuration.copy(
                frameProcessor = frameProcessor
        )
    }

    /**
     * @param frameProcessor receives preview frames for processing.
     * @see FrameProcessorJava
     */
    fun frameProcessor(frameProcessor: FrameProcessorJava?): FotoapparatBuilder = apply {
        configuration = configuration.copy(
                frameProcessor = frameProcessor?.let { it::process }
        )
    }

    /**
     * @param logger logger which will print logs. No logger is set by default.
     * @see io.fotoapparat.log.Loggers
     */
    fun logger(logger: Logger): FotoapparatBuilder =
            apply { this.logger = logger }

    /**
     * @param callback which will be notified when camera error happens in Fotoapparat.
     * @see CameraErrorListener
     */
    fun cameraErrorCallback(callback: CameraErrorListener): FotoapparatBuilder =
            apply { cameraErrorCallback = { callback.onError(it) } }

    /**
     * @param callback which will be notified when camera error happens in Fotoapparat.
     * @see CameraErrorListener
     */
    fun cameraErrorCallback(callback: CameraErrorCallback): FotoapparatBuilder =
            apply { cameraErrorCallback = callback }

    /**
     * @param renderer view which will draw the stream from the camera.
     * @see CameraView
     */
    fun into(renderer: CameraRenderer): FotoapparatBuilder =
            apply { this.renderer = renderer }

    /**
     * @param focusView view which will be used for touch to focus.
     * @see FocusView
     */
    fun focusView(focusView: FocusView): FotoapparatBuilder =
            apply { this.focusView = focusView }

    /**
     * @return set up instance of [Fotoapparat].
     * @throws IllegalStateException if some mandatory parameters are not specified.
     */
    fun build() = buildInternal(
            renderer = renderer
    )

    private fun buildInternal(
            renderer: CameraRenderer?
    ): Fotoapparat {
        if (renderer == null) {
            throw IllegalStateException("CameraRenderer is mandatory.")
        }

        return Fotoapparat(
                context = context,
                view = renderer,
                focusView = focusView,
                lensPosition = lensPositionSelector,
                cameraConfiguration = configuration,
                scaleType = scaleType,
                cameraErrorCallback = cameraErrorCallback,
                logger = logger
        )
    }

}